[ONNX][dynamo_export] Skip instance_norm decomp for export

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120866

Otherwise, instance_norm is decomposed into batch_norm with training set to True.
Downstream exporter has no way to figure out that training is actually not needed.
On the other hand, ONNX does have InstanceNormalization operator defined, however
due to decomp, it unnecessarily exports as batch norm and glue code.

Depends on https://github.com/microsoft/onnxscript/pull/1284

BowenBao committed:
[ONNX][dynamo_export] Skip instance_norm decomp for export

Otherwise, instance_norm is decomposed into batch_norm with training set to True.
Downstream exporter has no way to figure out that training is actually not needed.
On the other hand, ONNX does have InstanceNormalization operator defined, however
due to decomp, it unnecessarily exports as batch norm and glue code.

Depends on https://github.com/microsoft/onnxscript/pull/1284

[ghstack-poisoned]

BowenBao committed:
update unittest on "[ONNX][dynamo_export] Skip instance_norm decomp for export"


Otherwise, instance_norm is decomposed into batch_norm with training set to True.
Downstream exporter has no way to figure out that training is actually not needed.
On the other hand, ONNX does have InstanceNormalization operator defined, however
due to decomp, it unnecessarily exports as batch norm and glue code.

Depends on https://github.com/microsoft/onnxscript/pull/1284

[ghstack-poisoned]

BowenBao commented on test/onnx/test_fx_op_consistency.py:
```diff
@@ -1548,6 +1538,11 @@ def skip_torchlib_forward_compatibility(
     ),
+    xfail(
+        "nn.functional.instance_norm",
+        model_type=pytorch_test_common.TorchModelType.TORCH_EXPORT_EXPORTEDPROGRAM,
```
cc @titaiwangms, similar to what @liqunfu observed the other day, I need to move this xfail downwards to here to correctly only xfail on 'TORCH_EXPORT_EXPORTEDPROGRAM'.

titaiwangms replied:
Tracked by https://github.com/pytorch/pytorch/issues/120918.
I will use this to see the case.

titaiwangms replied:
Revisiting on this. I think with matcher seems reasonable to put it down here though.

titaiwangms replied:
I just confirmed that model_type only works with subfails.

thiagocrepaldi commented:
assuming that manipulating the decomposition table is not enough, we should discuss with Meta a mechanism to prevent decomposition to happen?

titaiwangms commented on test/onnx/test_fx_op_consistency.py:
```diff
@@ -1548,6 +1538,11 @@ def skip_torchlib_forward_compatibility(
+    xfail(
+        "nn.functional.instance_norm",
+        model_type=pytorch_test_common.TorchModelType.TORCH_EXPORT_EXPORTEDPROGRAM,
+        reason="fixme: Assertion error: result mismatch",
```
Is it because decomp slipping does not work with `ExportedProgram`, it fails?

BowenBao replied:
correct

BowenBao replied:
Well, to be more precise, decomp skipping were not enabled during `torch.export`. It is only enabled during `torch.onnx.dynamo_export`.

titaiwangms reviewed (approved):
Such a quick support!

BowenBao committed:
relax fp16 tolerance, forgot to include on "[ONNX][dynamo_export] Skip instance_norm decomp for export"


Otherwise, instance_norm is decomposed into batch_norm with training set to True.
Downstream exporter has no way to figure out that training is actually not needed.
On the other hand, ONNX does have InstanceNormalization operator defined, however
due to decomp, it unnecessarily exports as batch norm and glue code.

Depends on https://github.com/microsoft/onnxscript/pull/1284

[ghstack-poisoned]

BowenBao committed:
some export program subtest passes on "[ONNX][dynamo_export] Skip instance_norm decomp for export"


Otherwise, instance_norm is decomposed into batch_norm with training set to True.
Downstream exporter has no way to figure out that training is actually not needed.
On the other hand, ONNX does have InstanceNormalization operator defined, however
due to decomp, it unnecessarily exports as batch norm and glue code.

Depends on https://github.com/microsoft/onnxscript/pull/1284

[ghstack-poisoned]

BowenBao committed:
switch back to release onnxscript on "[ONNX][dynamo_export] Skip instance_norm decomp for export"


Otherwise, instance_norm is decomposed into batch_norm with training set to True.
Downstream exporter has no way to figure out that training is actually not needed.
On the other hand, ONNX does have InstanceNormalization operator defined, however
due to decomp, it unnecessarily exports as batch norm and glue code.

Depends on https://github.com/microsoft/onnxscript/pull/1284

[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge
